testing/base/source_gen.cpp

```diff
@@ -0,0 +1,676 @@
+
+auto SourceGen::Global() -> SourceGen& {
+  static SourceGen global_gen;
+  return global_gen;
+}
+
+SourceGen::SourceGen(Language language) : language_(language) {}
+
+// Some heuristic numbers used when formatting generated code.
+constexpr static int NumSingleLineFunctionParams = 3;
+constexpr static int NumSingleLineMethodParams = 2;
+constexpr static int MaxParamsPerLine = 4;
+
+static auto EstimateAvgFunctionDeclLines(SourceGen::FunctionDeclParams params)
+    -> double {
+  // Currently model a uniform distribution [0, max] parameters. Assume a line
+  // break before the first parameter for >3 and after every 4th.
+  int param_lines = 0;
+  for (int num_params : llvm::seq_inclusive(0, params.max_params)) {
+    if (num_params > NumSingleLineFunctionParams) {
+      param_lines += (num_params + MaxParamsPerLine - 1) / MaxParamsPerLine;
+    }
+  }
+  return 1.0 + static_cast<double>(param_lines) / (params.max_params + 1);
+}
+
+static auto EstimateAvgMethodDeclLines(SourceGen::MethodDeclParams params)
+    -> double {
+  // Currently model a uniform distribution [0, max] parameters. Assume a line
+  // break before the first parameter for >2 and after every 4th.
+  int param_lines = 0;
+  for (int num_params : llvm::seq_inclusive(0, params.max_params)) {
+    if (num_params > NumSingleLineMethodParams) {
+      param_lines += (num_params + MaxParamsPerLine - 1) / MaxParamsPerLine;
+    }
+  }
+  return 1.0 + static_cast<double>(param_lines) / (params.max_params + 1);
+}
+
+// Note that this should match the heuristics used when formatting.
+// TODO: See top-level TODO about line estimates and formatting.
+static auto EstimateAvgClassLines(SourceGen::ClassParams params) -> double {
+  // Blank line, comment line, and class open line.
+  double avg = 3.0;
+
+  // One comment line and blank line per function, plus the function lines.
+  avg +=
+      (2.0 + EstimateAvgFunctionDeclLines(params.public_function_decl_params)) *
+      params.public_function_decls;
+  avg += (2.0 + EstimateAvgMethodDeclLines(params.public_method_decl_params)) *
+         params.public_method_decls;
+  avg += (2.0 +
+          EstimateAvgFunctionDeclLines(params.private_function_decl_params)) *
+         params.private_function_decls;
+  avg += (2.0 + EstimateAvgMethodDeclLines(params.private_method_decl_params)) *
+         params.private_method_decls;
+
+  // A blank line and all the fields (if any).
+  if (params.private_field_decls > 0) {
+    avg += 1.0 + params.private_field_decls;
+  }
+
+  // No need to account for the class close line, we have an extra blank line
+  // count for the last of the above.
+  return avg;
+}
+
+auto SourceGen::GenAPIFileDenseDecls(int target_lines, DenseDeclParams params)
+    -> std::string {
+  std::string source;
+  llvm::raw_string_ostream os(source);
+
+  double avg_class_lines = EstimateAvgClassLines(params.class_params);
+  int num_classes = static_cast<double>(target_lines) / avg_class_lines;
```

Comment:
I think this should be
```suggestion
  CARBON_CHECK(target_lines >= 4);
  int num_classes = static_cast<double>(target_lines - 4) / avg_class_lines;
```
to account for the five lines of comment below, and the fact that one of the classes doesn't have its own blank line.

Reply:
Somewhat tweaked after changing the way we handle the estimation as suggested, and added a named constant for this. Should address the core idea though.

I've also made the check a bit more strict to ensure we get at least one class.